File: app/src/main/java/com/anbillon/lunarlite/di/module/DbModule.java
package com.anbillon.lunarlite.di.module;

import android.content.Context;
import android.database.sqlite.SQLiteOpenHelper;
import com.anbillon.lunarlite.data.db.AssetsDbOpenHelper;
import com.anbillon.lunarlite.data.db.datasource.AlmanacDataSource;
import com.anbillon.lunarlite.domain.repository.AlmanacRepository;
import com.squareup.sqlbrite.BriteDatabase;
import com.squareup.sqlbrite.SqlBrite;
import dagger.Module;
import dagger.Provides;
import javax.inject.Singleton;
import rx.schedulers.Schedulers;
import timber.log.Timber;

/**
 * Dagger module that provides database related collaborators.
 *
 * @author Vincent Cheung (coolingfall@gmail.com)
 */
@Module public final class DbModule {
	private static final String DB_NAME = "almanac.db";
	private static final int DB_VERSION = 1;

	@Provides @Singleton SQLiteOpenHelper provideDbOpenHelper(Context context) {
		return new AssetsDbOpenHelper(context, DB_NAME, DB_VERSION);
	}

	@Provides @Singleton SqlBrite provideSqlBrite() {
		return SqlBrite.create(message -> Timber.d(message));
	}

	@Provides @Singleton BriteDatabase provdeBriteDatabase(SqlBrite sqlBrite,
		SQLiteOpenHelper helper) {
		return sqlBrite.wrapDatabaseHelper(helper, Schedulers.io());
	}

	@Provides @Singleton AlmanacRepository provideResitory(AlmanacDataSource almanacDataSource) {
		return almanacDataSource;
	}
}

File: app/src/main/java/com/anbillon/lunarlite/di/module/DbModule.kt
package com.anbillon.lunarlite.di.module

import android.content.Context
import android.database.sqlite.SQLiteOpenHelper
import com.anbillon.lunarlite.data.db.AssetsDbOpenHelper
import com.anbillon.lunarlite.data.db.repository.AlmanacDataSource
import com.coolerfall.lunarlite.domain.repository.AlmanacRepository
import com.squareup.sqlbrite2.BriteDatabase
import com.squareup.sqlbrite2.SqlBrite
import dagger.Module
import dagger.Provides
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import javax.inject.Singleton

/**
 * @author Vincent Cheung (coolingfall@gmail.com)
 */
@Module
class DbModule {
  companion object {
    private val DB_NAME = "almanac.db"
    private val DB_VERSION = 1
  }

  @Provides
  @Singleton internal fun provdeBriteDatabase(context: Context): BriteDatabase {
    val helper: SQLiteOpenHelper = AssetsDbOpenHelper(context,
        DB_NAME,
        DB_VERSION)
    val sqlBrite = SqlBrite.Builder().logger { message -> Timber.d(message) }.build()
    return sqlBrite.wrapDatabaseHelper(helper, Schedulers.io())
  }

  @Provides
  @Singleton internal fun provideAlmanacRepository(
      dataSource: AlmanacDataSource): AlmanacRepository {
    return dataSource
  }
}